MINOR: Add varint serde utilities for new message format

hachikuji opened the pull request.

hachikuji committed:
MINOR: Add varint serde utilities for new message format

hachikuji commented:
cc @ijuma @junrao

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -340,61 +340,61 @@ static void checkException(IOException e, TestCloseable... closeablesWithExcepti
         int value2 = 0xf4f3f2f1;
         ByteArrayOutputStream os2 = new ByteArrayOutputStream();
-        Utils.writeUnsignedIntLE(os2, value2);
+        ByteUtils.writeUnsignedIntLE(os2, value2);
```
I think we should move these tests.

hachikuji replied:
Ack. Thought I looked for these, but apparently not too carefully 😉 .

hachikuji committed:
Move normal unsigned int serde test cases into ByteUtils

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+     * were used to encode this integer.
+     *
+     * @param in The input to read from
+     * @return The integer read (MUST BE TREATED WITH SPECIAL CARE TO AVOID SIGNEDNESS)
```
Since we are returning a `long`, is this warning needed?

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+     * @throws IllegalArgumentException if variable-length value does not terminate
+     *                                  after 5 bytes have been read
+     */
+    private static long readRawVarUnsignedLong(ByteBuffer buffer) {
```
I was checking the code in the protobuf repository and it turns out that the optimised version of this code is pretty complex:

https://github.com/google/protobuf/blob/master/java/core/src/main/java/com/google/protobuf/CodedInputStream.java#L2473

I think it makes sense to start with the simple implementation, but we should file a JIRA for considering the optimisations in their implementation.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+        return readRawUnsignedVarInt(in) & 0xffffffffL;
+    }
+
+    private static int readRawUnsignedVarInt(DataInput in) throws IOException {
```
It seems like we are not totally consistent in the naming convention: `readRawUnsignedVarInt` versus `readRawVarUnsignedInt` below.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+     * Read an unsigned long stored in variable-length format from
+     * <a href="http://code.google.com/apis/protocolbuffers/docs/encoding.html">
+     * Google Protocol Buffers</a>. Also update the index to indicate how many bytes
+     * were used to encode this long.
```
I think `update the index` could be made clearer since the `DataInput` interface doesn't mention an index. Same for other similar cases.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+     * Read an unsigned integer stored in variable-length format from
+     * <a href="http://code.google.com/apis/protocolbuffers/docs/encoding.html">
+     * Google Protocol Buffers</a>. Also update the index to indicate how many bytes
+     * were used to encode this integer.
```
We should probably say `update the buffer position` instead of `update index`. Same for other similar cases.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+     *
+     * @param longValue The unsigned integer represented as a long
+     */
+    public static int bytesForVarUnsignedIntEncoding(long longValue) {
```
What if the `longValue` is outside the expected range?

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/ByteUtilsTest.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    @Test
+    public void testVarIntSerde() {
```
A few more tests that I think would be useful (for documentation, verification and to make it easier to optimise in the future):

1. Boundary cases for 2, 3, 4, 6, 7, 8, 9 bytes (the latter 3 only apply for longs).

2. Exceptional conditions: we throw an IllegalArgumentException in some cases, we should also exercise cases where the data type we pass allows for out of range values (unsigned cases are probably the most common case).

3. It would be good to have some tests that verify the buffer content (like the non varint tests in the class). Without that, there is the potential for regressions where we can still roundtrip, but we don't follow the spec any more. One way to do that would be to add a third parameter to the `assert*` methods.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+                throw new IllegalArgumentException("Variable length quantity is too long");
+            }
+        }
+        return value | (b << i);
```
Hmm, the slow path protobuf code seems simpler:

```java
      long result = 0;
      for (int shift = 0; shift < 64; shift += 7) {
        final byte b = readRawByte();
        result |= (long) (b & 0x7F) << shift;
        if ((b & 0x80) == 0) {
          return result;
        }
      }
      throw InvalidProtocolBufferException.malformedVarint();
```

Is the code in PR does this way for performance reasons?

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+    private static long readRawVarUnsignedLong(DataInput in) throws IOException {
+        long value = 0L;
+        int i = 0;
+        long b;
```
Can this be a `byte`?

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+    private static int readRawVarUnsignedInt(ByteBuffer buffer) {
+        int value = 0;
+        int i = 0;
+        int b;
```
Can this be a `byte`?

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+    private static int readRawUnsignedVarInt(DataInput in) throws IOException {
+        int value = 0;
+        int i = 0;
+        int b;
```
Can this be a `byte`.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+    private static long readRawVarUnsignedLong(ByteBuffer buffer) {
+        long value = 0L;
+        int i = 0;
+        long b;
```
Can this be a `byte`?

hachikuji replied:
Yes, seems so.

hachikuji replied:
On second thought, I think it is probably better to leave this as a long. We need to cast to a long before shifting anyway, so I'm not sure there is any benefit.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+     */
+    public static long readVarLong(DataInput in) throws IOException {
+        long raw = readRawVarUnsignedLong(in);
+        // This undoes the trick in writeSignedVarLong()
```
We don't have `writeSignedVarLong`. It's mentioned in at least one other place as well. The protobuf code doesn't seem to do this. Do we need to?

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,454 @@
+            buffer.put(b);
+            value >>>= 7;
+        }
+        buffer.put((byte) (value & 0x7f));
```
Is the `& 0x7f` needed here? The protobuf code doesn't seem to do that:

```java
          if ((value & ~0x7F) == 0) {
            buffer[position++] = (byte) value;
            totalBytesWritten++;
            return;
          } else {
            buffer[position++] = (byte) ((value & 0x7F) | 0x80);
            totalBytesWritten++;
            value >>>= 7;
          }
```

hachikuji replied:
I think you are right. Breaking from the loop ensures that the upper nibble of the last byte is already 0, so the `& 0x7f` should be redundant.

ijuma reviewed (commented):
Thanks for the PR. Left a few comments and questions.

hachikuji committed:
Address review comments

hachikuji committed:
Move ByteBufferInputStream and ByteBufferOutputStream into utils

hachikuji commented:
@ijuma This is ready for another look. Several changes:

1. As discussed offline, I've removed support for the unsigned encoding.
2. I've moved `ByteBufferInputStream` and `ByteBufferOutputStream` into `utils`. I did the same thing in the message format patch, but thought I might as well do it here so that I didn't have to mess with the import control file in order to use these classes for testing.
3. A few minor improvements to the serdes themselves plus the additional test cases. I actually did find a bug handling one of the error cases, so one more win for unit tests!

hachikuji committed:
Use less verbose name for byte count utility

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java:
```diff
@@ -473,4 +474,60 @@ public Records validate(Object item) {
+
+        @Override
+        public Integer read(ByteBuffer buffer) {
+            return ByteUtils.readVarint(buffer);
```
Looks like a copy/paste issue. It should return a `Long` and call `readVarlong`. Should we have some simple tests for the `Type` instances?

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,327 @@
+     *
+     * @param value The signed value
+     */
+    public static int sizeOfVarint(int value) {
```
This is a security risk because the right bit pattern can cause an infinite loop. We need to stop after a certain size. Same for `sizeOfVarlong`.

ijuma replied:
My bad, the parameter is just an `int` and not a buffer. Thanks to @hachikuji for correcting me.

ijuma reviewed (commented):
Thanks for the updates. I spotted a couple of potential issues, which should be easy to fix.

Aside from that, I put together some changes that are more subjective:

https://github.com/ijuma/kafka/commit/28cf239a53fa3e620266ab8cbc2c79d0f08805da

Please include the ones you agree with and we can discuss the ones that you don't.

hachikuji committed:
Various improvements and fixes

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -0,0 +1,324 @@
+    public static void writeVarlong(long value, ByteBuffer buffer) {
+        long v = (value << 1) ^ (value >> 63);
+        while ((v & 0xffffffffffffff80L) != 0L) {
+            byte b = (byte) (((int) v & 0x7f) | 0x80);
```
Do you need this `int` cast? I thought I removed that in my branch and it worked fine.

hachikuji replied:
Ah, must have missed that.

ijuma reviewed (approved):
Thanks for the updates. Nice idea to use variables that look like hex values to avoid the casts in the tests.

Just one comment, LGTM otherwise.

hachikuji committed:
Remove unneeded integer cast

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -258,7 +258,7 @@ public static void writeVarint(int value, ByteBuffer buffer) {
         long v = (value << 1) ^ (value >> 63);
         while ((v & 0xffffffffffffff80L) != 0L) {
-            out.writeByte(((int) v & 0x7f) | 0x80);
+            out.writeByte((byte) (v & 0x7f) | 0x80);
```
Since the type expected by `writeByte` is `int`, it seems like the previous code was OK? Otherwise we do two conversions.
